plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts

```diff
@@ -326,42 +331,32 @@ export class DefaultEntitiesCatalog implements EntitiesCatalog {
   }
 
   async facets(request: EntityFacetsRequest): Promise<EntityFacetsResponse> {
-    const { entities } = await this.entities({
-      filter: request.filter,
-      authorizationToken: request.authorizationToken,
-    });
-
     const facets: EntityFacetsResponse['facets'] = {};
+    const db = this.database;
 
     for (const facet of request.facets) {
-      const values = entities
-        .map(entity => {
-          // TODO(freben): Generalize this code to handle any field that may
-          // have dots in its key?
-          if (facet.startsWith('metadata.annotations.')) {
-            return entity.metadata.annotations?.[
-              facet.substring('metadata.annotations.'.length)
-            ];
-          } else if (facet.startsWith('metadata.labels.')) {
-            return entity.metadata.labels?.[
-              facet.substring('metadata.labels.'.length)
-            ];
-          }
-          return lodash.get(entity, facet);
-        })
-        .flatMap(field => {
-          if (typeof field === 'string') {
-            return [field];
-          } else if (Array.isArray(field)) {
-            return field.filter(i => typeof i === 'string');
-          }
-          return [];
+      const dbQuery = db('search')
+        .join('final_entities', 'search.entity_id', 'final_entities.entity_id')
+        .where('search.key', facet)
+        .count('search.entity_id as count')
+        .select({
+          value: 'search.original_value',
         })
-        .sort();
+        .groupBy('search.original_value');
+
+      if (request?.filter) {
+        parseFilter(request.filter, dbQuery, db, false, 'search.entity_id');
+      }
 
-      const counts = lodash.countBy(values, lodash.identity);
+      const result = await Promise.all([dbQuery]);
 
-      facets[facet] = Object.entries(counts).map(([value, count]) => ({
+      const finalResult: Record<string, any> = {};
+      result.map(entries => {
+        entries.map(data => {
+          finalResult[data.value] = data.count;
+        });
+      });
+      facets[facet] = Object.entries(finalResult).map(([value, count]) => ({
```

Comment:
I don't quite understand these multiple loops. Why can't the result be gotten immediately from the output of the query? Should be possible to essentially do `facets[facet] = result.map(r => ({ value: r.value, count: r.count }));` shouldn't it